orttraining/orttraining/training_api/onnxruntime_training_c_api.cc

```diff
@@ -210,16 +216,34 @@ ORT_API_STATUS_IMPL(OrtTrainingApis::OptimizerStep, _Inout_ OrtTrainingSession*
   API_IMPL_END
 }
 
-ORT_API_STATUS_IMPL(OrtTrainingApis::RegisterLinearLRScheduler, _Inout_ OrtTrainingSession* sess,
-                    _In_ int64_t warmup_step_count, _In_ int64_t total_step_count) {
+ORT_API_STATUS_IMPL(OrtTrainingApis::RegisterLRScheduler, _Inout_ OrtTrainingSession* sess,
+                    _In_ void* lr_scheduler_parameters, _In_ enum OrtLRSchedulerType lr_scheduler_type,
+                    _In_opt_ const float* initial_lr) {
   API_IMPL_BEGIN
 
+  OrtStatus* status = nullptr;
+
   auto session = reinterpret_cast<onnxruntime::training::api::TrainingSession*>(sess);
-  ORT_API_RETURN_IF_STATUS_NOT_OK(session->RegisterScheduler([warmup_step_count, total_step_count](auto optimizer) {
-    return std::make_unique<onnxruntime::training::api::LinearLRScheduler>(optimizer, warmup_step_count, total_step_count);
-  }));
+  switch (lr_scheduler_type) {
+    case OrtLRSchedulerType::LinearLRScheduler: {
+      auto parameters = reinterpret_cast<OrtLinearLRSchedulerParameters*>(lr_scheduler_parameters);
```

Comment:
I think you need also check whether parameters is nullptr or now (considering user probabbly give a cosine_schduler_parameters).

Reply:
Oh Okay,  I thought we are using dynamic_cast. I feel it is dangerous here we probably cast a cosine parameter pointer  to  linear parameter pointer without check?

Reply:
dynamic_cast might be useful for up casts and down casts. In this case, we want to cast the pointer back to the original type based on the enum. We rely on the user to provide the correct enum.